### PR TITLE
Speed up HTTP/2 header iteration

### PR DESCRIPTION
**Pull request by chrisvest**
Motivation:
When iterating HTTP/2 headers, the pseudo-headers must always come first in iteration order.
We previously solved this by iterating the headers twice, filtering, and concatenating the two filtered iterations.
This was very slow, but meant we could use the MultiMap pretty much unchanged.

Modification:
Add another pair of linkage references to the entries in the MultiMap.
This doubly-linked list preserves total insertion order, except pseudo-headers are always arranged before normal headers.
We use a handful of fields to keep track of where pseudo-headers start and end, and likewise for normal headers.
For iteration we then use this new chain instead of the bucket-head and entries chains.

Result:
Iteration of HTTP/2 headers is now over 700% faster, but 'add' and 'addAll' methods are slower by 22-44%, when benchmarked with 30 headers.

### CODE REVIEW

**Comment by chrisvest** [comment]
Visualization of the benchmark results:
image

**Review by Scottmitch — APPROVED** [review]
approach lgtm, good work @chrisvest! note I didn't look closely at the linking details but I assume you have this covered with tests.

File: codec-http2/src/main/java/io/netty5/handler/codec/http2/headers/DefaultHttp2Headers.java [hunk 1]
```diff
@@ -262,69 +268,138 @@ public CharSequence status() {
...
+                }
+            } else {
+                entry.linkAfter(lastNormalHeader);
+                lastNormalHeader = entry;
+            }
+        }
+        return entry;
+    }
+
+    private static class Http2MultiMapEntry extends MultiMapEntry<CharSequence, CharSequence> {
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty5/handler/codec/http2/headers/DefaultHttp2Headers.java:323]:
```suggestion
    private static final class Http2MultiMapEntry extends MultiMapEntry<CharSequence, CharSequence> {
```

File: codec-http2/src/main/java/io/netty5/handler/codec/http2/headers/DefaultHttp2Headers.java [hunk 2]
```diff
@@ -262,69 +268,138 @@ public CharSequence status() {
...
+    @Override
+    public Iterator<Entry<CharSequence, CharSequence>> iterator() {
+        return isEmpty() ? emptyIterator() : new FullHttp2EntryIterator(
+                firstPseudoHeader != null ? firstPseudoHeader : firstNormalHeader);
     }
 
-    private static <T> Iterator<T> filter(Iterator<T> iterator, Predicate<T> predicate) {
-        if (!iterator.hasNext()) {
-            return iterator;
+    private class FullHttp2EntryIterator implements Iterator<Entry<CharSequence, CharSequence>> {
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty5/handler/codec/http2/headers/DefaultHttp2Headers.java:372]:
```suggestion
    private final class FullHttp2EntryIterator implements Iterator<Entry<CharSequence, CharSequence>> {
```

**Review by normanmaurer — CHANGES_REQUESTED** [review]
two small things... Great work!

**Pull request merged**